File: ruby/task_queue_example.rb
require 'sinatra'
require_relative 'task_queue'

WORKLOAD = 3..5

configure do
  set :server_settings, max_threads: 5
  set :queue, TaskQueue.start
end

task_queue = 'ok'

helpers do
  def queue
    settings.queue
  end

  def heavy_work
    sleep rand(WORKLOAD)
    logger.info('done')
  end
end

get '/' do
  queue.push(-> { heavy_work })
end

File: elixir/task_queue_example/lib/task_queue_example.ex
defmodule TaskQueueExample do
  @moduledoc false

  use Plug.Router

  require Logger

  alias TaskQueueExample.TaskQueue

  plug Plug.Logger
  plug :match
  plug :dispatch

  @workload 3..5

  get "/" do
    TaskQueue.push(&heavy_work/0)
    send_resp(conn, :ok, "")
  end

  defp heavy_work do
    timeout = Enum.random(@workload) * 1_000
    Process.sleep(timeout)
    Logger.info("done")
  end
end
